libs/ballot-interpreter-nh-next/src/geometry.rs

```diff
@@ -0,0 +1,402 @@
+            let normalized_again = super::normalize_angle(normalized);
+            assert_nearly_eq!(normalized, normalized_again);
+        }
+
+        #[test]
+        fn prop_normalize_angle_is_equivalent(angle in ANGLE_RANGE) {
+            let normalized = super::normalize_angle(angle);
+            let equivalent = super::normalize_angle(angle + PI);
+            assert_nearly_eq!(normalized, equivalent);
+        }
+    }
+}
+
+#[cfg(test)]
+mod normalize_center_of_rect {
+    use proptest::prelude::*;
+
+    #[test]
+    fn test_center_of_rect() {
+        let rect = super::Rect::new(0, 0, 10, 10);
+        let center = rect.center();
+        assert_eq!(center.x, 4.5);
+        assert_eq!(center.y, 4.5);
+    }
+
+    #[test]
+    fn test_center_of_rect_with_odd_dimensions() {
+        let rect = super::Rect::new(0, 0, 11, 11);
+        let center = rect.center();
+        assert_eq!(center.x, 5.0);
+        assert_eq!(center.y, 5.0);
+    }
+
+    proptest! {
+        #[test]
+        fn prop_center_of_rect_is_in_rect(x in 0i32..100i32, y in 0i32..100i32, width in 1u32..100u32, height in 1u32..100u32) {
+            let rect = super::Rect::new(x, y, width, height);
+            let center = rect.center();
+            prop_assert!((rect.left() as f32) <= center.x);
+            prop_assert!(center.x <= (rect.right() as f32));
+            prop_assert!((rect.top() as f32) <= center.y);
+            prop_assert!(center.y <= (rect.bottom() as f32));
+        }
+    }
+}
+
+pub fn find_best_line_through_items(rects: &Vec<Rect>, angle: f32, tolerance: f32) -> Vec<Rect> {
+    if rects.is_empty() {
+        return vec![];
+    }
+
+    let best_rects: Vec<&Rect> = rects
+        .par_iter()
+        .fold_with(vec![], |best_rects, rect| {
+            let mut best_rects = best_rects;
+
+            for other_rect in rects.iter() {
+                let rect_center = rect.center();
+                let other_rect_center = other_rect.center();
+                let line_angle = (other_rect_center.y - rect_center.y)
+                    .atan2(other_rect_center.x - rect_center.x);
+
+                if angle_diff(line_angle, angle) > tolerance {
+                    continue;
+                }
+
+                let rects_intsersecting_line = rects
+                    .iter()
+                    .filter(|r| {
+                        rect_intersects_line(r, &Segment::new(rect_center, other_rect_center))
+                    })
+                    .collect::<Vec<&Rect>>();
+
+                if rects_intsersecting_line.len() > best_rects.len() {
+                    best_rects = rects_intsersecting_line;
+                }
+            }
+
+            best_rects
+        })
+        .reduce_with(|best_rects, other_best_rects| {
+            if other_best_rects.len() > best_rects.len() {
+                other_best_rects
+            } else {
+                best_rects
+            }
+        })
```

Comment:
I'm having a hard time understanding the flow of this algorithm. Maybe we can talk it through?